From: Salil <salil.mehta@huawei.com>
Date: Fri, 18 Aug 2017 12:31:37 +0100
Subject: net: hns3: Fixes the missing u64_stats_fetch_begin_irq in 64-bit
 stats fetch
Patch-mainline: v4.14-rc1
Git-commit: d36d36ceac11e514f1c313ea6ca7568e2463b6f2
References: bsc#1104353 FATE#326415

This patch fixes the missing u64_stats_fetch_begin_irq() while trying to
atomically do 64-bit RX/TX fetch. We did not get any error during test
as our SoC is 64-bit so all of these seq/lock operations results in NOOP.

As such, this seq lock supports has been added for the sake of completion
if this code ever runs on 32-bit platform and we are trying to do 64-bit
stats fetch.

Fixes: 76ad4f0ee747 ("net: hns3: Add support of HNS3 Ethernet Driver for
hip08 SoC")
Reported-by: Dan Carpenter <dan.carpenter@oracle.com>
Signed-off-by: Salil Mehta <salil.mehta@huawei.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/net/ethernet/hisilicon/hns3/hns3pf/hns3_enet.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/net/ethernet/hisilicon/hns3/hns3pf/hns3_enet.c
+++ b/drivers/net/ethernet/hisilicon/hns3/hns3pf/hns3_enet.c
@@ -1054,6 +1054,7 @@ hns3_nic_get_stats64(struct net_device *
 		/* fetch the tx stats */
 		ring = priv->ring_data[idx].ring;
 		do {
+			start = u64_stats_fetch_begin_irq(&ring->syncp);
 			tx_bytes += ring->stats.tx_bytes;
 			tx_pkts += ring->stats.tx_pkts;
 		} while (u64_stats_fetch_retry_irq(&ring->syncp, start));
@@ -1061,6 +1062,7 @@ hns3_nic_get_stats64(struct net_device *
 		/* fetch the rx stats */
 		ring = priv->ring_data[idx + queue_num].ring;
 		do {
+			start = u64_stats_fetch_begin_irq(&ring->syncp);
 			rx_bytes += ring->stats.rx_bytes;
 			rx_pkts += ring->stats.rx_pkts;
 		} while (u64_stats_fetch_retry_irq(&ring->syncp, start));
